dids/diddht/diddht.go

```diff
@@ -1,175 +1,191 @@
+	"github.com/tbd54566975/web5-go/dids/diddht/internal/pkarr"
+	"github.com/tv42/zbase32"
 )
 
-// Decoder is used to structure the DNS representation of a DID
-type Decoder struct {
-	rootRecord string
-	records    map[string]string
+// relay is the internal interface used to publish Pakrr messages to the DHT
+type relay interface {
+	Put(didID string, payload *bep44.Message) error
+	PutWithContext(ctx context.Context, didID string, payload *bep44.Message) error
+
+	Fetch(didID string) (*bep44.Message, error)
+	FetchWithContext(ctx context.Context, didID string) (*bep44.Message, error)
 }
 
-func (rec *Decoder) DIDDocument(didURI string) (*didcore.Document, error) {
-	if len(rec.rootRecord) == 0 {
-		return nil, errors.New("no root record found")
-	}
-	relationshipMap, err := parseVerificationRelationships(rec.rootRecord)
-	if err != nil {
-		return nil, err
-	}
+var defaultRelay relay
+var once sync.Once
 
-	// Now we have a did in a dns record. yay
-	document := &didcore.Document{
-		ID: didURI,
-	}
+// getDefaultRelay returns the default Pkarr relay client.
+func getDefaultRelay() relay {
+	once.Do(func() {
+		defaultRelay = pkarr.NewPkarrRelay("", http.DefaultClient)
+	})
 
-	// Now create the did document
-	for name, data := range rec.records {
-		switch {
-		case strings.HasPrefix(name, "_k"):
-			var vMethod didcore.VerificationMethod
-			if err := UnmarshalVerificationMethod(data, &vMethod); err != nil {
-				// TODO handle error
-				continue
-			}
+	return defaultRelay
+}
 
-			// extracting kN from _kN._did
-			entryID := strings.Split(name, ".")[0][1:]
-			relationships, ok := relationshipMap[entryID]
+// dHTDidOptions is used to configure the creation of a `did:dht` DID; it's an internal representation of the options
+type dHTDidOptions struct {
+	algorithmID         string
+	keyManager          crypto.KeyManager
+	services            []*didcore.Service
+	verificationMethods []didcore.VerificationMethod
+	purposes            map[string][]didcore.Purpose
+	relay               relay
+}
 
-			if !ok {
-				// no relationships
-				continue
-			}
+type DHTDidOption func(o *dHTDidOptions)
 
-			opts := []didcore.Purpose{}
-			for _, r := range relationships {
-				if o, ok := vmPurposeDNStoDID[r]; ok {
-					opts = append(opts, didcore.Purpose(o))
-				}
-			}
+// WithServices adds the provided services to the DID document.
+func WithServices(services ...*didcore.Service) DHTDidOption {
+	return func(o *dHTDidOptions) {
+		o.services = append(o.services, services...)
+	}
+}
 
-			document.AddVerificationMethod(
-				vMethod,
-				didcore.Purposes(opts...),
-			)
-		case strings.HasPrefix(name, "_s"):
-			var s didcore.Service
-			if err := UnmarshalService(data, &s); err != nil {
-				// TODO handle error
-				continue
+// WithVerificationMethod adds the provided verification method to the DID document.
+func WithVerificationMethod(method didcore.VerificationMethod, purposes []didcore.Purpose) DHTDidOption {
```

Comment:
https://github.com/TBD54566975/web5-go/pull/33/files#r1492018221

Reply:
hmm what usage pattern did you have in mind here? the goal is to provide the caller with the ability to specify:
* the algorithm id to use in order to generate the private key
* what purpose the key serves

[here](https://github.com/TBD54566975/web5-go/blob/dht-create/dids/didweb/didweb.go#L56-L69) is what i did for `did:web`

Reply:
again whichever route we go with, we should make consistent across different did implementations

I think I'm more sympathetic w/ @mistermoe here b/c `WithVerificationMethod` comes off to me as the caller is defining the full VM but (at least?) part of that (the public key) is generated by the creation process... I don't love `PrivateKeys` in that it reads like the caller is passing in a full private key... maybe we should just atomize it into two different functions `WithAlgorithmID` and `WithPurposes` (following the `With` pattern)

Reply:
I was thinking of what made me go this route. My hypothesis was that `diddht` itself will be an underlying "implementation detail" of say the dids package. Right now `dids` only has resolve, I believe. in that case, having diddht be more configurable seemed sane to me. 

I think this may have been wrong; while `resolve` is based on a uri that includes the method and it can then use the respective method implementation, I guess create is not the same, and the consumers of the sdk would know specifically what method they want to use, it wouldn't be deduced from a uri.

Reply:
@mistermoe skimmed through the `did:web` create options and they look very similar to what I'd use here. I'm thinking we can push them up in the stack and reuse. Thoughts?